gateway-messages/src/sp_impl.rs

```diff
@@ -493,11 +506,31 @@ pub fn handle_message<H: SpHandler>(
             )
         }
         Some(OutgoingTrailingData::CabooseData(data)) => {
-            out[n..n + data.len()].copy_from_slice(data);
-            data.len()
+            // Because we didn't know the encoded length of the packet until
+            // now, we have to call back into the handler and copy caboose data
+            // into the trailing region.
+            //
+            // Unfortunately, unlike every other possible form of trailing data,
+            // this is faillible!  We check for errors here and re-serialize the
+            // whole chunk of packet data if there's a failure.
+            let len = data.len();
+            if let Err(e) =
+                handler.copy_caboose_value_into(data, &mut out[n..][..len])
+            {
+                let response = Message {
+                    header: Header { version: version::CURRENT, message_id },
+                    kind: MessageKind::SpResponse(SpResponse::Error(e)),
+                };
+
+                // We know `response` is well-formed and fits into `out`, so we
+                // can unwrap serialization.
+                n = hubpack::serialize(&mut out[..], &response).unwrap();
+            } else {
+                n += len;
```

Comment:
Tiny nit - I have a mild preference for the "success" case coming first. How do you think this would look as a `match` instead of `if let ... else ...`? (My preference for a `match` might be stronger if `copy_caboose_value_into` returned a length that we needed to add to `n`.)

Reply:
Sure, done in a09ea3743d8727e87512ebd86e248d7ff3c5484c